src/ivfs/ivfs_remote.ml

```diff
@@ -33,62 +51,94 @@ module Make (Store : Ivfs_tree.STORE) = struct
+  let mk_remote ?(url="") name repo =
     let session = Vfs.File.Stream.session None in
     let url_file, remote_url = mk_url url in
     let head_file, update_head = mk_head session in
-    let t = { remote_url; update_head } in
-    let fetch_file = mk_fetch t make_task repo in
+    let fetch, fetch_file = mk_fetch ~remote_url ~update_head repo in
     let files = [
-      Vfs.Inode.file "url" url_file;
-      Vfs.Inode.file "head" head_file;
+      Vfs.Inode.file "url"   url_file;
+      Vfs.Inode.file "head"  head_file;
       Vfs.Inode.file "fetch" fetch_file;
     ] in
-    Vfs.Dir.of_list (fun () -> files)
+    let s = { remote_name = name; remote_url; update_head; fetch } in
+    s, Vfs.Dir.of_list (fun () -> files)
+
+  let has_name name (s, _) = s.remote_name = name
+  let find_inode name t = List.find (has_name name) t.state |> snd
 
-  let create ?(init=[]) make_task repo =
+  let inode_of_remote t ?url name =
+    try find_inode name t
+    with Not_found ->
+      let s, files = mk_remote ?url name t.repo in
+      let inode = Vfs.Inode.dir name files in
+      t.state <- (s, inode) :: t.state;
+      inode
+
+  let create ?(init=[]) repo =
     Log.debug (fun l -> l "create");
-    let remote ?url name =
-      Vfs.Inode.dir name (mk_remote ?url  make_task repo)
-    in
-    let init = List.map (fun (name, url) -> name, remote ~url name) init in
-    let remotes = ref init in
-    let ls () = Vfs.ok (List.map snd !remotes) in
+    let t = { root = Vfs.Dir.of_list (fun () -> []); state = []; repo } in
+    let ls () = Vfs.ok (List.map snd t.state) in
     let lookup n =
-      try Vfs.ok (List.assoc n !remotes)
+      try Vfs.ok (find_inode n t)
       with Not_found -> Vfs.Dir.err_no_entry
     in
     let mkdir n =
-      if List.mem_assoc n !remotes then Vfs.Dir.err_already_exists
-      else
-        let i = remote n in
-        remotes := (n, i) :: !remotes;
-        Vfs.ok ( i)
+      if List.exists (has_name n) t.state then Vfs.Dir.err_already_exists
+      else Vfs.ok (inode_of_remote t n)
     in
     let remove _ = Vfs.Dir.err_dir_only in
     let rename i new_name =
       let old_name = Vfs.Inode.basename i in
       Vfs.Inode.set_basename i new_name;
-      let rs = List.filter (fun (n,_) -> n <>old_name) !remotes in
-      remotes := (new_name, i) :: rs;
+      t.state <- List.map (fun (s, i as x) ->
+          if s.remote_name = old_name then { s with remote_name = new_name }, i
+          else x
+        ) t.state;
       Vfs.ok ()
     in
-    Vfs.Dir.dir_only ~ls ~mkdir ~remove ~rename ~lookup
+    let root = Vfs.Dir.dir_only ~ls ~mkdir ~remove ~rename ~lookup in
+    List.iter (fun {name; url} ->
+        let (_:Vfs.Inode.t) = inode_of_remote t ?url name in ()
+      ) init;
+    t.root <- root;
+    t
+
+  let root t = t.root
+
+  let add t { name; url } =
+    t.state <- List.filter (fun s -> not @@ has_name name s) t.state;
+    let (_:Vfs.Inode.t) = inode_of_remote t ?url name in
+    ()
```

Comment:
This doesn't match with the documentation string. We currently cannot easily change the remote url from an existing remote, so I'd probably change that code to follow what the doc says.